### PR TITLE
ENH: Database Script

### LINKED ISSUE #1: Database
Creating utils for database script

### PR DESCRIPTION
**Pull request by MaharshiChoksi**
Merging database script to main branch

### CODE REVIEW

File: src/utils.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
+import mysql.connector as sql # from dotenv import load_dotenv
+import os
+from typing import Tuple
+from dotenv import load_dotenv, find_dotenv
+from mysql.connector import Error
```
Comment (mitulds) [inline src/utils.py]: sort using `isort`

File: src/utils.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+load_dotenv(override=True)
+
+# TODO-1: MySQL connection Method
+def connect_to_mysql(host:str='localhost', user:str='root', port:int=3306, database:str=None) -> Tuple[sql.connection.MySQLConnection, sql.connection.MySQLCursor]:
+    """ Connection to MySQL Server
```
Comment (mitulds) [inline src/utils.py]: make sure to put args on newline so the whole code is visible on the laptop screen when reviewing

File: src/utils.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        port (int, optional): port Number. Defaults to 3306.
+        database (str, optional): Database Name to connect at login. Defaults to None.
+
+    Returns:
+        Tuple[sql.connection.MySQLConnection, sql.connection.MySQLCursor]: returns sql connection string and cursor in form of tuple
+    """
+
+    try:
+        connection = sql.connect(host=host, user=user, port=port, password=os.environ['PASSWORD'], database=database)
+
```
Comment (mitulds) [inline src/utils.py]: remove empty line

File: src/utils.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
+import mysql.connector as sql # from dotenv import load_dotenv
+import os
+from typing import Tuple
+from dotenv import load_dotenv, find_dotenv
+from mysql.connector import Error
+
+load_dotenv(override=True)
+
+# TODO-1: MySQL connection Method
+def connect_to_mysql(host:str='localhost', user:str='root', port:int=3306, database:str=None) -> Tuple[sql.connection.MySQLConnection, sql.connection.MySQLCursor]:
```
Comment (mitulds) [inline src/utils.py]: don't need to define default values for the args, only keep the database value

File: src/utils.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        connection = sql.connect(host=host, user=user, port=port, password=os.environ['PASSWORD'], database=database)
+
+        cursor = connection.cursor()
+        return connection, cursor
+    except Error as conn_err:
+        raise(conn_err)
+
+
+# TODO-2: Database Creation Method
+def create_database(conn:sql.connection.MySQLConnection, cursor:sql.connection.MySQLCursor) -> bool :
```
Comment (mitulds) [inline src/utils.py]: create a new args called db_name and pass into your function

File: src/utils.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    Returns:
+        bool: returns True/False if query executed successfully or not
+    """
+    
+    try:
+        query_drop_database= "DROP DATABASE IF EXISTS ds_practice;"
+        query_create_database= "CREATE DATABASE ds_practice;"
+        cursor.execute(query_drop_database)
+        cursor.execute(query_create_database)
+        conn.commit()
```
Comment (mitulds) [inline src/utils.py]: only use `commit` when sending data to the table

File: src/utils.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+        print("Database Created")
+        return True
```
Comment (mitulds) [inline src/utils.py]:
```
cur.execute("SHOW DATABASES;")
databases = cur.fetchall()
return databases
```

File: src/utils.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    except Error as db_create_err:
+        raise(db_create_err)
+
+
+# TODO-3: Table Creation Method
+def create_table(conn:sql.connection.MySQLConnection, cursor:sql.connection.MySQLCursor, database:str) -> bool:
```
Comment (mitulds) [inline src/utils.py]: create a new args for table_name

File: src/utils.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    try:
+        query_db = f"USE {database};"
+        query_drop_table = "DROP TABLE IF EXISTS students;"
+        query_create_table = "CREATE TABLE students(id int NOT NULL, first_name varchar(255) NOT NULL, last_name varchar(255) NOT NULL, age int NOT NULL, phone bigint);"
+        cursor.execute(query_db)
+        cursor.execute(query_drop_table)
+        cursor.execute(query_create_table)
+        conn.commit()
+        print("Table Created")
+        return True
```
Comment (mitulds) [inline src/utils.py]: dont need `commit` and `return`

File: src/utils.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+        bool: returns True if connection is closed else False
+    """
+    
+    try:
+        curs.close()
+        conn.close()
+        print("Connection Closed")
+        return True
+    except Error as conn_error:
+        raise(conn_error)
```
Comment (mitulds) [inline src/utils.py]: dont need this function

File: src/utils.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+    except Error as error:
+        conn.rollback()
+        print(error)
+    finally:
+        close_connection(curs=curs, conn=conn)
+
+
+if __name__ == '__main__':
+    main()
```
Comment (mitulds) [inline src/utils.py]: create a new script `database.py`

**Review by MaharshiChoksi — COMMENTED** [review]
Updated code with changes

File: requirements.txt [hunk 12]
```diff
@@ -0,0 +1,5 @@
+os
+typing
```
Comment (mitulds) [inline requirements.txt]: built-in so dont need here

File: src/database.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
+from utils import *
```
Comment (mitulds) [inline src/database.py]: only import modules that are been used in the script

File: src/utils.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+def get_data_from_csv(fname:str) -> pd.DataFrame:
+    """Reading csv file and returning Dataframe
+
+    Args:
+        fname (str): location where csv file is located
+
+    Returns:
+        _type_: return pandas datatype
+    """
+    df = pd.read_csv(fname)
```
Comment (mitulds) [inline src/utils.py:159]: add new line `df.drop(['Unnamed: 0'], axis=1, inplace=True, errors='ignore')`

**Review by mitulds — APPROVED** [review]

**Pull request merged**